src/sentry/replays/lib/selector/parse.py

```diff
@@ -0,0 +1,97 @@
+from typing import List, Optional, Union
+
+from cssselect import Selector, SelectorSyntaxError
+from cssselect import parse as cssselect_parse
+from cssselect.parser import Attrib, Class, CombinedSelector, Element, Hash
+from rest_framework.exceptions import ParseError
+
+SelectorType = Union[Attrib, Class, Element, Hash]
+
+
+class QueryType:
+    def __init__(self):
+        self.alt: Optional[str] = None
+        self.aria_label: Optional[str] = None
+        self.classes: List[str] = []
+        self.id: Optional[str] = None
+        self.role: Optional[str] = None
+        self.tag: Optional[str] = None
+        self.testid: Optional[str] = None
+        self.title: Optional[str] = None
+
+
+def parse_selector(css_selector: str) -> List[QueryType]:
+    try:
+        selectors: List[Selector] = cssselect_parse(css_selector)
+    except SelectorSyntaxError:
+        # Invalid selector syntax. No query data can be extracted.
+        return []
+
+    queries: List[QueryType] = []
+    for selector in selectors:
+        query = QueryType()
+        visit_selector_tree(query, selector.parsed_tree)
+        queries.append(query)
+    return queries
+
+
+def visit_selector_tree(query: QueryType, selector: SelectorType) -> None:
+    """Visit selector tree ignoring unhandled items.
+
+    We intentionally ignore specificity and psuedo-elements.
+
+    Cssselect refers to a selector as a "parsed_tree". While this is true its best thought of as
+    a linked-list. The first element encountered in the list is the last condition defined in the
+    selector text. The "Element" class is always the tail and has no children.
+
+    For example:
+
+        Attrib -> Class -> Hash -> Class -> Attrib -> Element
+    """
+    if isinstance(selector, Attrib):
+        visit_attribute(query, selector)
+        visit_selector_tree(query, selector.selector)
+    elif isinstance(selector, Class):
+        visit_class(query, selector)
+        visit_selector_tree(query, selector.selector)
+    elif isinstance(selector, Element):
+        visit_element(query, selector)
+        return None
+    elif isinstance(selector, Hash):
+        visit_hash(query, selector)
+        visit_selector_tree(query, selector.selector)
+    elif isinstance(selector, CombinedSelector):
+        raise ParseError("Nested selectors are not supported.")
+    else:
+        # We ignore unhandled selector types rather than erroring.
```

Comment:
are we actually ignoring unhandled selector types? it seems like we just continue to visit them based on this code?

Reply:
Yeah this would recurse the children assuming the outer wrapper did not matter.  That's a bad design so I'm erring on these unsupported selector types.